go/internal/registry/schema.go

```diff
@@ -0,0 +1,236 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package registry
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/google/dotprompt/go/dotprompt"
+	"github.com/invopop/jsonschema"
+	orderedmap "github.com/wk8/go-ordered-map/v2"
+)
+
+// DefineSchema registers a Go struct as a schema with the given name.
+func (r *Registry) DefineSchema(name string, structType any) error {
+	jsonSchema, err := convertStructToJsonSchema(structType)
+	if err != nil {
+		return err
+	}
+
+	if r.Dotprompt == nil {
+		r.Dotprompt = dotprompt.NewDotprompt(&dotprompt.DotpromptOptions{
+			Schemas: map[string]*jsonschema.Schema{},
+		})
+	}
+
+	r.Dotprompt.DefineSchema(name, jsonSchema)
+	r.RegisterValue(SchemaType+"/"+name, structType)
+	fmt.Printf("Registered schema '%s' with registry and Dotprompt\n", name)
+	return nil
+}
+
+// RegisterSchemaWithDotprompt registers a schema with the Dotprompt instance
+// This is used during Init to register schemas that were defined before the registry was created.
+func (r *Registry) RegisterSchemaWithDotprompt(name string, schema any) error {
+	if r.Dotprompt == nil {
+		r.Dotprompt = dotprompt.NewDotprompt(&dotprompt.DotpromptOptions{
+			Schemas: map[string]*jsonschema.Schema{},
+		})
+	}
+
+	jsonSchema, err := convertStructToJsonSchema(schema)
+	if err != nil {
+		return err
+	}
+
+	r.Dotprompt.DefineSchema(name, jsonSchema)
+	r.RegisterValue(SchemaType+"/"+name, schema)
+
+	// Set up schema lookup if not already done
+	r.setupSchemaLookupFunction()
+
+	return nil
+}
+
+// setupSchemaLookupFunction registers the external schema lookup function with Dotprompt
+// This function bridges between Dotprompt's schema resolution and the registry's values
+func (r *Registry) setupSchemaLookupFunction() {
+	if r.Dotprompt == nil {
+		return
+	}
+
+	r.Dotprompt.RegisterExternalSchemaLookup(func(schemaName string) any {
+		schemaValue := r.LookupValue(SchemaType + "/" + schemaName)
+		if schemaValue != nil {
+			return schemaValue
+		}
+		return nil
+	})
+}
+
+// DumpRegistrySchemas prints all schemas stored in the registry
+func (r *Registry) DumpRegistrySchemas() {
+	fmt.Println("=== Registry Schemas ===")
+
+	for k, v := range r.values {
+		if strings.HasPrefix(k, SchemaType+"/") {
+			schemaName := strings.TrimPrefix(k, SchemaType+"/")
+			fmt.Printf("Schema: %s, Type: %T\n", schemaName, v)
+		}
+	}
+
+	fmt.Println("=======================")
+}
+
+// convertStructToJsonSchema converts a Go struct to a JSON schema
+func convertStructToJsonSchema(structType any) (*jsonschema.Schema, error) {
```

Comment:
Wouldn't it be better if we use `map[string]any` as the type instead of just any?